### PR TITLE
Fix: rework clusters to align velaux and cli

### LINKED ISSUE #3238: Failed to connect cluster from cloud, but secret is created
**Describe the bug**

Failed to connect cluster from the cloud in Velux, but `vela cluster list` show:
```
ack                                    X509Certificate https://123.56.27.128:6443 
c1c25ad5a841b4b0        X509Certificate https://123.56.175.182:6443
c1cf869cc45f6439         X509Certificate https://47.242.126.78:6443 
c5ac2f1160b9747b          X509Certificate https://123.57.73.107:6443
```

**To Reproduce**
The user does not have permission, so the cluster connect fail.

**Screenshots**
image

**KubeVela Version**
1.2.3

### PR DESCRIPTION
**Pull request by Somefive**
Signed-off-by: Somefive <yd219913@alibaba-inc.com>

### Description of your changes

1. Use **VirtualCluster** struct to unifies the differences of implementations for cluster secret and OCM managed cluster.
2. Align Vela CLI and VelaUX to use the same underlying cluster management operation functions. (Except that current `cluster join` for OCM is closely coupled with CLI operations, this should be reworked later)
3. Enhance `vela cluster list` which show labels now. (`vela cluster label` command should be implemented to help user manage the labels of clusters.)

Fixes #3238 

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: pkg/multicluster/cluster_management.go [hunk 1]
```diff
@@ -17,174 +17,416 @@ limitations under the License.
...
+		_, _ = fmt.Fprintf(&clusterConfig.Logs, "failed to parse server endpoint: %v", err)
+	}
+	return clusterConfig, nil
+}
+
+const (
+	// ClusterGateWayEngine cluster-gateway cluster management solution
+	ClusterGateWayEngine = "cluster-gateway"
+	// OCMEngine ocm cluster management solution
+	OCMEngine = "ocm"
```
Comment (wonderflow) [inline pkg/multicluster/cluster_management.go:264]: what's the difference between these two engines? AFAIK, using OCM we also need cluster-gateway.

Comment (Somefive) [reply pkg/multicluster/cluster_management.go:264]: For `cluster-gateway` (by default), the implementation of `cluster` is a secret. For `ocm`, the implementation of `cluster` is OCM ManagedCluster. The join and detach logic will differ a lot. The name of `engine` here is a legacy design by @yue9944882 . Normally, we do not need to care about them, except when users want to use OCM.

Comment (wonderflow) [reply pkg/multicluster/cluster_management.go:264]: OK

File: pkg/multicluster/suite_test.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2021 The KubeVela Authors.
```
Comment (wonderflow) [inline pkg/multicluster/suite_test.go:2]:
```suggestion
Copyright 2022 The KubeVela Authors.
```

Comment (Somefive) [reply pkg/multicluster/suite_test.go]: Several files changed from 2021 to 2021-2022.

**Review by wonderflow — APPROVED** [review]

File: pkg/multicluster/cluster_management.go [hunk 3]
```diff
@@ -17,174 +17,416 @@ limitations under the License.
...
-	if !errors2.IsNotFound(err) {
-		return errors.Wrapf(err, "failed to check duplicate cluster secret")
+	if err := cli.Create(ctx, secret); err != nil {
+		return errors.Wrapf(err, "failed to add cluster to kubernetes")
+	}
+	// TODO(somefive): create namespace now only work for cluster secret
+	if clusterConfig.CreateNamespace != "" {
+		if err := ensureNamespaceExists(ctx, cli, clusterConfig.ClusterName, clusterConfig.CreateNamespace); err != nil {
+			_ = cli.Delete(ctx, secret)
+			return errors.Wrapf(err, "failed to ensure vela-system namespace installed in cluster %s", clusterConfig.ClusterName)
```
Comment (barnettZQG) [inline pkg/multicluster/cluster_management.go:121]:
```suggestion
			return errors.Wrapf(err, "failed to ensure %s namespace installed in cluster %s",  clusterConfig.CreateNamespace, clusterConfig.ClusterName)
```

File: pkg/multicluster/o11n.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	var workerNumber, masterNumber int
+	var memoryCapacity, cpuCapacity, podCapacity, memoryAllocatable, cpuAllocatable, podAllocatable resource.Quantity
+	for _, node := range nodes.Items {
+		if _, ok := node.Labels["node-role.kubernetes.io/master"]; ok {
+			masterNumber++
+		} else {
+			workerNumber++
+		}
+		capacity := node.Status.Capacity
+		memoryCapacity.Add(*capacity.Memory())
```
Comment (barnettZQG) [inline pkg/multicluster/o11n.go:59]: Ignore the not schedulable node?

File: pkg/apiserver/rest/usecase/cluster.go [hunk 5]
```diff
@@ -198,14 +199,14 @@ func joinClusterByKubeConfigString(ctx context.Context, k8sClient client.Client,
 	defer func() {
 		_ = os.Remove(tmpFileName)
 	}()
-	cluster, err := multicluster.JoinClusterByKubeConfig(ctx, k8sClient, tmpFileName, clusterName)
+	clusterConfig, err := multicluster.JoinClusterByKubeConfig(ctx, k8sClient, tmpFileName, clusterName, multicluster.JoinClusterCreateNamespaceOption(velatypes.DefaultKubeVelaNS))
 	if err != nil {
 		if errors.Is(err, multicluster.ErrClusterExists) {
 			return "", bcode.ErrClusterExistsInKubernetes
```
[the remark this answers is not in the document]
Comment (Somefive) [reply pkg/apiserver/rest/usecase/cluster.go:205]: The failure of this error can be different. We can handle limited scenarios such as ClusterExists. We can add more error handlers if we have better choice.

File: pkg/multicluster/o11n.go (hunk 4 above)
Comment (Somefive) [reply pkg/multicluster/o11n.go:59]: Currently not. The unschedulable node might only be temporarily unavailable. The total capacity should count it. We have allocatable cpu/memory metrics and they should take node schedulability into account. This interface is previously designed for apiserver only. We are working on a newer architecture for the multi-cluster observability. This file will be extended soon.

File: pkg/multicluster/cluster_management.go (hunk 3 above)
Comment (Somefive) [reply pkg/multicluster/cluster_management.go:121]: Good idea.

File: pkg/multicluster/o11n.go (hunk 4 above)
Comment (barnettZQG) [reply pkg/multicluster/o11n.go:59]: Temporary solutions are ok.

File: pkg/apiserver/rest/usecase/cluster.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Somefive) [reply pkg/apiserver/rest/usecase/cluster.go:205]: The sync problem between cluster created by CLI and cluster created by VelaUX will be resolved in another PR.

**Review by barnettZQG — APPROVED** [review]

**Review by yangsoon — APPROVED** [review]

**Review by zzxwill — APPROVED** [review]

**Review by leejanee — APPROVED** [review]

**Pull request merged**